api/service/project_service.go

```diff
@@ -1,49 +1,132 @@
-// Copyright 2020 The Merlin Authors
-//
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at
-//
-//      http://www.apache.org/licenses/LICENSE-2.0
-//
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
 package service
 
 import (
 	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	cache "github.com/patrickmn/go-cache"
 
-	"github.com/caraml-dev/merlin/mlp"
+	mlp "github.com/caraml-dev/merlin/mlp"
+	mlpClient "github.com/caraml-dev/mlp/api/client"
 )
 
+const (
+	projectCacheExpirySeconds  = 600
+	projectCacheCleanUpSeconds = 900
+	projectCacheKeyPrefix      = "proj:"
+
+	// mlpQueryTimeoutSeconds is the timeout that will be used on the API calls to the MLP server.
+	mlpQueryTimeoutSeconds = 30
+)
+
+// ProjectsService provides a set of methods to interact with the MLP / Merlin APIs
 type ProjectsService interface {
-	List(ctx context.Context, projectName string) (mlp.Projects, error)
+	// List lists available projects, optionally filtered by given project `name`
+	List(ctx context.Context, name string) (mlp.Projects, error)
+	// GetByID gets the project matching the provided id.
 	GetByID(ctx context.Context, projectID int32) (mlp.Project, error)
-	GetByName(ctx context.Context, projectName string) (mlp.Project, error)
 }
 
-func NewProjectsService(mlpAPIClient mlp.APIClient) ProjectsService {
-	return &projectsService{
-		mlpAPIClient: mlpAPIClient,
+type projectsService struct {
+	mlpClient mlp.APIClient
+	cache     *cache.Cache
+}
+
+// NewProjectsService returns a service that retrieves information that is shared across MLP projects.
+func NewProjectsService(mlpAPIClient mlp.APIClient) (ProjectsService, error) {
+	svc := &projectsService{
+		mlpClient: mlpAPIClient,
+		cache:     cache.New(projectCacheExpirySeconds*time.Second, projectCacheCleanUpSeconds*time.Second),
```

Comment:
The implementation is lgtm.

However, can we re-use `pkg/transformer/cache` that uses `github.com/coocood/freecache` instead of introducing new dependency (`github.com/patrickmn/go-cache`)? If needed, we can refactor it to `pkg/cache`.